### PR TITLE
Fix Django path normalization, test against Django 2.2

### LINKED ISSUE #122: Path normalization in a Django app that uses `include` falis with an exception
If the `urlpatterns` for an app uses `include` to bring in a `URLconf` that contains parameters, an attempt to normalize the path of a matching request will raise an exception.

For example, the `urlpatterns` are written this way in the [chipy.org](https://github.com/chicagopython/chipy.org) app: https://github.com/chicagopython/chipy.org/blob/main/chipy_org/urls.py#L16 . This issue causes many tests to fail, including this one: https://github.com/chicagopython/chipy.org/blob/main/chipy_org/apps/job_board/tests/test_views.py#L80 .

### PR DESCRIPTION
**Pull request by apotterri**
Fixes #122.

### CODE REVIEW

File: appmap/django.py [hunk 1]
```diff
@@ -133,25 +133,32 @@ def request_params(request):
...
+        # No kwargs mean no parameters to normalize
+        return path_info
+
+    # Get the current resolver. Look through its cache of reverse
+    # route mappings and find the regex that matches the current
+    # request. The route mappings are tuples of the form:
+    #   (format string, regex, args, kwargs)
+    resolver = get_resolver()
+    regex = None
+    for _,regex,_,_ in resolver.reverse_dict.values():
```
Comment (dividedmind) [inline appmap/django.py:146]: `resolver.reverse_dict` is a multi valued dict, so `values()` will skip some items if there are views which match more than one pattern. If you use `user_view` instead of a separate `included_user_view` in the test below you can reproduce this issue.

File: appmap/django.py [hunk 2]
```diff
@@ -133,25 +133,32 @@ def request_params(request):
...
-    # internal parser to turn it into a regex.
-    regex = resolved.route
-    match = re.match(regex, path_info[1:])
-    if not match:
-        regex = _route_to_regex(resolved.route)[0]
+    if not resolved.kwargs:
+        # No kwargs mean no parameters to normalize
+        return path_info
+
+    # Get the current resolver. Look through its cache of reverse
```
Comment (dividedmind) [inline appmap/django.py:140]: Is there a way to use `resolved.tried[-1]` (which should be the pattern that ended up matching) instead of trying to duplicate what `URLResolver.resolve` is doing?

File: appmap/test/test_django.py [hunk 3]
```diff
@@ -99,23 +99,36 @@ def user_post_view(_request, username, post_id):
...
-    django.urls.path('echo', echo_view),
+    path('test', view),
+    path('', view),
+    re_path('^user/(?P<username>[^/]+)$', user_view),
+    path('post/<int:post_id>', post_view),
+    path('post/<username>/<int:post_id>/summary', user_post_view),
+    re_path(r'^post/unnamed/(\d+)$', post_unnamed_view),
+    path('echo', echo_view),
+    re_path(r'^post/(included)/', include([
+            path('<username>', user_included_view)]))
```
Comment (dividedmind) [inline appmap/test/test_django.py:114]: As mentioned above, if you use `user_view` here you'll catch a potential problem with the normalization code.

File: appmap/test/test_django.py [hunk 4]
```diff
@@ -99,23 +99,36 @@ def user_post_view(_request, username, post_id):
...
+    assert appmap.enabled()     # sanity check
     # unnamed captures in a re_path don't show up in the event's
     # message attribute.
     assert len(events[0].message) == 0
 
+@pytest.mark.appmap_enabled
+def test_included_view(client, events):
+    client.get('/post/included/test_user')
+
+    assert len(events) > 0
```
Comment (dividedmind) [inline appmap/test/test_django.py:131]: Shouldn't this verify that the normalized path info ends up being `post/included/{username}`?

File: appmap/test/test_django.py [hunk 5]
```diff
@@ -99,23 +99,36 @@ def user_post_view(_request, username, post_id):
...
-    django.urls.re_path(r'^post/unnamed/(\d+)$', post_unnamed_view),
-    django.urls.path('echo', echo_view),
+    path('test', view),
+    path('', view),
+    re_path('^user/(?P<username>[^/]+)$', user_view),
+    path('post/<int:post_id>', post_view),
+    path('post/<username>/<int:post_id>/summary', user_post_view),
+    re_path(r'^post/unnamed/(\d+)$', post_unnamed_view),
+    path('echo', echo_view),
+    re_path(r'^post/(included)/', include([
```
Comment (dividedmind) [inline appmap/test/test_django.py:113]: Why the parentheses?

**Review by dividedmind — CHANGES_REQUESTED** [review]
I've found some corner cases where this approach can fail.

File: appmap/django.py (hunk 2 above)
Comment (apotterri) [reply appmap/django.py:140]: No, because `resolved.tried` was introduced in 3.2.

File: appmap/test/test_django.py (hunk 4 above)
Comment (apotterri) [reply appmap/test/test_django.py:131]: At the very least, yes. Thanks.

File: appmap/test/test_django.py (hunk 5 above)
Comment (apotterri) [reply appmap/test/test_django.py:113]: I believe they're leftover from some debugging. I'll double-check. If they are actually accomplishing something, I'll add a comment to indicate what. Otherwise, I'll remove them.

File: appmap/django.py (hunk 2 above)
Comment (dividedmind) [reply appmap/django.py:140]: Ah, ok, thanks. If there isn't any other way then I suppose this is acceptable, I just hope it isn't too brittle.

File: appmap/django.py (hunk 1 above)
Comment (apotterri) [reply appmap/django.py:146]: Ok, thanks, I'll investigate.

File: appmap/django.py (hunk 2 above)
Comment (apotterri) [reply appmap/django.py:140]: We could see if `hasattr(resolved, 'tried')` and use it if it's available, falling back to using `reverse_dict` if it's not. The 2.2 implementation seems unlikely to change at this point, and `tried` is part of the 3.2 interface. 

Does the added complexity seem worth it to you?

File: appmap/test/test_django.py (hunk 5 above)
Comment (apotterri) [reply appmap/test/test_django.py]: Didn't need them, took them out.

File: appmap/test/test_django.py (hunk 3 above)
Comment (apotterri) [reply appmap/test/test_django.py]: Got it, thanks.

File: appmap/django.py (hunk 2 above)
Comment (apotterri) [reply appmap/django.py]: Well, it turns out using `tried` wasn't an option, after all. It's content for included paths isn't helpful. So, I settled on a single implementation.

File: pylintrc [hunk 6]
```diff
@@ -284,7 +284,7 @@ argument-naming-style=snake_case
 
 # Regular expression matching correct argument names. Overrides argument-
 # naming-style.
-#argument-rgx=
+argument-rgx=(?P<exempt>[a-z_]{1,2})
```
Comment (dividedmind) [inline pylintrc:287]: Not sure about this. I believe there's a good reason for the rule rejecting short names in general. They're fine for very short methods (or for contexts and lambda expressions, in which cases they are already allowed), but in general the scope of a name can be long enough that a short name unnecessarily makes the code less clear. (Same is true for variable names, below).

**Review by dividedmind — APPROVED** [review]
Thanks, looks fine now.

But please consider reverting the short argument and variable names exemption before merging. I don't think it's a good idea.
